controller/src/main/java/io/pravega/controller/store/stream/records/RecordHelper.java

```diff
@@ -0,0 +1,172 @@
+                    null;
+            return segmentsToSeal.stream().allMatch(segmentId -> {
+                if (isManualScale) {
+                    // compare segmentNumbers
+                    return segmentNumbersToSeal.contains(StreamSegmentNameUtils.getSegmentNumber(segmentId));
+                } else {
+                    // compare segmentIds
+                    return record.getSegmentsToSeal().contains(segmentId);
+                }
+            });
+        }
+        return false;
+    }
+
+    /**
+     * Method to compute epoch transition record. It takes segments to seal and new ranges and all the tables and
+     * computes the next epoch transition record.
+     * @param currentEpoch current epoch record
+     * @param segmentsToSeal segments to seal
+     * @param newRanges new ranges
+     * @param scaleTimestamp scale time
+     * @return new epoch transition record based on supplied input
+     */
+    public static EpochTransitionRecord computeEpochTransition(EpochRecord currentEpoch, Set<Long> segmentsToSeal,
+                                                               List<Map.Entry<Double, Double>> newRanges, long scaleTimestamp) {
+        Preconditions.checkState(segmentsToSeal.stream().allMatch(currentEpoch::containsSegment), "Invalid epoch transition request");
+
+        int newEpoch = currentEpoch.getEpoch() + 1;
+        int nextSegmentNumber = currentEpoch.getSegments().stream().mapToInt(StreamSegmentRecord::getSegmentNumber).max().getAsInt() + 1;
+        Map<Long, Map.Entry<Double, Double>> newSegments = new HashMap<>();
+        for (int i = 0; i < newRanges.size(); i++) {
+            newSegments.put(computeSegmentId(nextSegmentNumber + i, newEpoch), newRanges.get(i));
+        }
+        return new EpochTransitionRecord(currentEpoch.getEpoch(), scaleTimestamp, segmentsToSeal, newSegments);
+
+    }
+    // endregion
+
+    /**
+     * Helper method to compute list of continuous ranges. For example, two neighbouring key ranges where,
+     * range1.high == range2.low then they are considered neighbours.
+     * This method reduces input range into distinct continuous blocks.
+     * @param input list of key ranges.
+     * @return reduced list of key ranges.
+     */
+    private static List<Map.Entry<Double, Double>> reduce(List<Map.Entry<Double, Double>> input) {
+        List<Map.Entry<Double, Double>> ranges = new ArrayList<>(input);
+        ranges.sort(Comparator.comparingDouble(Map.Entry::getKey));
+        List<Map.Entry<Double, Double>> result = new ArrayList<>();
+        double low = -1.0;
+        double high = -1.0;
+        for (Map.Entry<Double, Double> range : ranges) {
+            if (high < range.getKey()) {
+                // add previous result and start a new result if prev.high is less than next.low
+                if (low != -1.0 && high != -1.0) {
+                    result.add(new AbstractMap.SimpleEntry<>(low, high));
```

Comment:
Shouldn't we be adding `<high, range.getKey()>` instead? I guess I'm confused about what this logic is trying to achieve.

Reply:
The method does following:
sort the input range by key. 
Then it loops over the range and merges two adjacent ranges if they share a boundary.. 
If they are disjoint or if they overlap then ranges are not merged and included separately.